cpp/util/CurlDescPool.cpp

```diff
@@ -0,0 +1,426 @@
+      curlDescPool.m_lockSharedShare.unlock();
+  }
+
+  /**
+   * Get sub-pool associated to the specified URL
+   *
+   * @param url
+   *   url we are accessing
+   *
+   * @param isMulti
+   *   true if we are using multi curl descriptors
+   */
+  CurlDescPool::SubPool &CurlDescPool::getSubPool(const SFURL &url)
+  {
+    std::string endPointName = url.scheme() + "://" + url.authority();
+
+    if (url.isProxyEnabled())
+    {
+      const Snowflake::Client::Util::Proxy & proxy = url.getProxy();
+      if (proxy.getMachine().empty())
+      {
+        endPointName += " no proxy";
+      }
+      else
+      {
+        endPointName += " with Proxy: " + proxy.getHost() +
+                        ", port: " + std::to_string(proxy.getPort()) +
+                        ", no_proxy: " + proxy.getNoProxy();
+      }
+    }
+
+    // iterator to lookup sub pools in our map
+    SubPoolByName_t::iterator subPoolIterator;
+
+    // see if sub pool exists, if not create it
+    std::lock_guard<std::mutex> guard(m_lockPool);
+
+    // find this sub pool
+    subPoolIterator = m_subPools.find(endPointName);
+
+    // sub-pool to return
+    SubPool *subPool;
+
+    // add it if it does not exists
+    if (subPoolIterator == m_subPools.end())
+    {
+      std::unique_ptr<SubPool> subPoolPt(new SubPool(endPointName,
+                                                     m_curlShared,
+                                                     *this));
+      subPool = subPoolPt.get();
+      m_subPools[endPointName] = std::move(subPoolPt);
+    }
+    else
+      subPool = subPoolIterator->second.get();
+
+    // trace
+    CXX_LOG_TRACE("CurlDescPool::getSubPool(): Getting subpool for endPoint=%s res=%p",
+                  endPointName.c_str(), (void *)subPool);       
+
+    // return sub pool
+    return(*subPool);
+  }
+
+  /**
+   * Constructor
+   *
+   * @param endPointName
+   *   name of the server end point, derived from the URL
+   *
+   * @param (IN/NULL) curlShareDesc
+   *   curl shared descriptor to use if non null
+   */
+  CurlDescPool::SubPool::SubPool(const std::string &endPointName,
+                                 CURLSH *curlShareDesc,
+                                 CurlDescPool& parentPool)
+    : m_endPointName(endPointName)
+    , m_curlShareDesc(curlShareDesc)
+    , m_parentPool(parentPool)
+  {
+  }
+
+  /**
+   * Get new curl descriptor from the pool and populate unique_ptr with it
+   *
+   * @param curlDescPt
+   *   empty unique pointer we will populate with a new pointer to a curl
+   *   descriptor
+   */
+  void CurlDescPool::SubPool::newCurlDesc(std::unique_ptr<CurlDesc> &curlDescPt)
+  {
+    // ensure not already created
+    if (!curlDescPt)
+    {
+      // get mutex and grab an entry from the subpool
+      bool isEmpty;
+      {
+        std::lock_guard<std::mutex> mutexGuard(m_lockSubPool);
+
+        // if pool empty, allocate new one
+        isEmpty = m_curlDescPool.empty();
+        if (isEmpty)
```

Comment:
could we add the braces even for the if body when are single line?

Reply:
sure. It's copied from odbc I didn't check the coding style when doing the copy.